mem/mem.py

```diff
@@ -0,0 +1,80 @@
+from common.configurable_model import ConfigurableModel
+import functools
+from bit_vector import BitVector
+from enum import Enum
+
+
+class Mode(Enum):
+    LINE_BUFFER = 0
+    FIFO = 1
+    SRAM = 2
+
+
+def check_addr(fn):
+    @functools.wraps(fn)
+    def wrapped(self, addr, *args):
+        if not (0 <= addr < self.data_depth):
+            raise ValueError(f"Invalid address ({addr}),"
+                             f" must be within range(0, {self.data_depth})")
+        return fn(self, addr, *args)
+    return wrapped
+
+
+def gen_mem(data_width: int,
+            data_depth: int):
+    class Mem(ConfigurableModel(32, 32)):
```

Comment:
Looking at this further, I think the correct interface is as follows:

- `__call__()` emulates the big `always` block in `memory_core.vp` https://github.com/rsetaluri/magma_cgra/blob/8c7905f871e77ac3982c05fddb43f67200c5397c/mem/genesis/memory_core.vp#L172. We can leverage `kwargs` so that we don't immediately bloat the model with all the unused things like `chain_in` and `chain_out`. But I think it would be good to throw in the `wen_in` and `ren_in` signals and test that logic as well.
- `__read()` and `__write()` can be private helper functions.